examples/graphrnn_s.py

```diff
@@ -0,0 +1,362 @@
+import argparse
+import random
+
+import matplotlib.pyplot as plt
+import networkx as nx
+import torch
+from torch import nn
+from torch.nn import functional as F
+from torch.nn.utils import rnn as rnnutils
+
+import torch_geometric
+from torch_geometric import data, loader
+from torch_geometric import transforms as T
+
+
+class EncodeGraphRNNFeature(T.BaseTransform):
+    r"""Encode the graph adjacency matrix into a sequence of adjacency vectors
+    of length :obj:`M`.
+    This is the approach described in the "GraphRNN: Generating Realistic
+    Graphs with Deep Auto-regressive Models"
+    <https://arxiv.org/pdf/1802.08773>`_ paper.
+
+    Args:
+        M (int): The length of the adjacency vectors.
+    """
+    def __init__(self, M):
+        self.M = M
+
+    @staticmethod
+    def extract_bands(adj, M):
+        r"""Uses stride tricks to extract :obj:`M`-long bands above the
+        diagonal of the given square matrix.
+
+        Args:
+            adj (Tensor): The adjacency matrix of shape :obj:`(N, N)`.
+            M (int): The number of bands to extract.
+
+        :rtype: Tensor
+        """
+        N = adj.shape[1]
+        adj = adj.reshape(N, N)
+        padded_adj = torch.zeros((N + M - 1, N))
+        padded_adj[M - 1:, :] = adj
+        return padded_adj.as_strided(size=(N - 1, M), stride=(N + 1, N),
+                                     storage_offset=1)
+
+    @staticmethod
+    def bands_to_matrix(bands):
+        r"""Given M bands above the diagonal of a square matrix, return the
+        full matrix.
+
+        Args:
+            bands (Tensor): The M bands of shape :obj:`(N, M)`.
+
+        :rtype: Tensor
+        """
+        M = bands.shape[1]
+        N = bands.shape[0] + 1
+        padded_adj = torch.zeros((N + M - 1, N))
+        view = padded_adj.as_strided(size=(N - 1, M), stride=(N + 1, N),
+                                     storage_offset=1)
+        view[:, :] = bands
+        return padded_adj[M - 1:, :]
+
+    @staticmethod
+    def inverse(y):
+        r"""Inverse of the :obj:`__call__` method: given the encoded sequence
+        y, return the adjacency matrix.
+
+        Args:
+            y (Tensor): The encoded sequence of shape :obj:`(N, M)`.
+
+        :rtype: Tensor
+        """
+        bands = torch.flip(y, dims=[1])
+        adj = EncodeGraphRNNFeature.bands_to_matrix(bands)
+        return adj
+
+    def __call__(self, data):
+        adj = torch_geometric.utils.to_dense_adj(data.edge_index)
```

Comment:
Is it possible to do this without going to the dense adjacency matrix? Perhaps it should be quite easy to extract the bands from a CSR format as sparse tensors?

Reply:
Yes, there are several options here:
Using a sparse mask (which is then used with `csr.multiply(mask).todia()` or `csr[mask]`)
```python
def construct_sparse_mask(N, M):
    """
    Construct a sparse mask consisting of the M diagonals above the main
    diagonal of a NxN matrix.
    """

    sum_to_M = int((M * (M + 1)) / 2)
    num_entries = N * M - sum_to_M
    data = np.ones((num_entries, ), dtype=np.bool_)

    start_x = np.repeat(np.arange(N - M), M)
    end_x = np.concatenate([np.full((i, ), N - i -1) for i in reversed(range(M))])
    indices_x = np.concatenate([start_x, end_x])

    ending_y = np.concatenate([np.arange(N - i, N) for i in reversed(range(M))])
    start_y = np.tile(np.arange(M) + 1, N - M)
    view_start_y = start_y.reshape(-1, M)
    view_start_y += np.arange(N - M).reshape(-1, 1)
    indices_y = np.concatenate([start_y, ending_y])

    return sp.coo_matrix((data, (indices_x, indices_y)), shape=(N, N))
```

Using diagonals:
```python
def extract_bands_sparse_dia(adj_sparse, M):
    diags = [np.pad(adj_sparse.diagonal(k=i), (i - 1, 0), mode='constant')
             for i in range(1, M + 1)]
    return np.stack(diags).T
```

However, it's a bit unclear what the best option is. For M << N `extract_bands_sparse_dia` is clearly faster, but when the graph is quite full (and `M` is high as a result, this may no longer be the case).

Any thoughts?

Reply:
Thanks for the detailed ideas! 

If I understand correctly, all three methods will essentially create a dense matrix anyway - in the first method above you create a dense mask, in the second the call `adj_sparse.diagonal` creates dense vectors, so if M is large it'll still be pretty large.

I can't see a big advantage between any of the approaches, except maybe to me the `extract_bands_sparse_dia` is most readable (but I agree it could be slow. In terms of performance, we might have to rely on the sampled graph being small enough to easily fit in memory as a dense graph?

What do you think?